Convert to an Ember CLI addon

azirbel opened the pull request:
This upgrade allows easy installs into Ember CLI apps, and is also fully
backwards-compatible with globals-based Ember apps.

* Convert from Coffeescript to JS
* Restructure to use Ember CLI conventions
* Blueprint to import Ember Table dependencies
* New demo app
    * Example code is auto-built from files backing the demo
* Add globals build, with build code in the `packaging` folder
* Fix/change to LazyContainerView involving `childViews`

See http://github.com/Addepar/ember-table-addon for more information.

Reviewers: @Addepar/ui2-developers especially @igillis @Dylnuge @cyril-sf 

This commit replaces current `ember-table` code with the `ember-table-addon` code. I recommend that you not read the diff in this PR directly, but instead look at http://github.com/Addepar/ember-table-addon.

Dylnuge commented:
:heart:

Dylnuge commented on .editorconfig:
```diff
@@ -0,0 +1,33 @@
+
+[*.html]
+indent_style = space
+indent_size = 2
```
Just curious, do these all need to be repeated if they're already rules under `*`?

Dylnuge commented on .travis.yml:
```diff
@@ -0,0 +1,20 @@
+    - node_modules
+
+before_install:
+  - "npm config set spin false"
```
Why quotes here but not in commands on lines 15-17 or 20?

azirbel replied:
dunno, autogenerated by ember cli

Dylnuge commented on Brocfile.js:
```diff
@@ -0,0 +1,35 @@
+var EmberAddon = require('ember-cli/lib/broccoli/ember-addon');
+var fs = require('fs');
+
+// TODO(azirbel): Use something more legit
```
I think JQuery has a method that can do this for you? This is alright as is but you might want to add a rule for single quote (replace `'` with `&#39;`)

Dylnuge commented on Brocfile.js:
```diff
@@ -0,0 +1,35 @@
+    patterns: [{
+      match: /@@{[^}]*}/g,
+      replacement: function(matchedText) {
+        filename = matchedText.slice(3, -1);
```
I suppose this is fine, but can't you capture the inner text using a regex group and consume that? IMHO that would be cleaner/more readable code.

cyril-sf replied:
I would personally in favor of removing the demo app from this repo

igillis replied:
@cyril-sf we need some easy way for folks to sanity check their changes. 

Maybe eventually that means having a separate demo app repo that you can `link` to this addon or something, but for now I think we should leave it here so you can just `ember server`.

cyril-sf replied:
I would assume that the test suite and their own app should be enough.

azirbel replied:
@Dylnuge like, `matchedText` would be the captured inner text? I don't think it's possible with the broccoli-replace addon I'm using, unfortunately. I could replace line 24 with a second regex, but I don't think that's helpful.

Dylnuge commented on Gruntfile.js:
```diff
@@ -3,194 +3,47 @@ module.exports = function (grunt) {
     banner: '/*!\n* <%=pkg.name %> v<%=pkg.version%>\n' +
             '* Copyright 2012-<%=grunt.template.today("yyyy")%> Addepar Inc.\n' +
-            '* See LICENSE.\n*/',
+            '* See LICENSE.md.\n*/',
```
Super nitpicky but I'd vote to remove the trailing `.`. This isn't a complete sentence anyways.

Dylnuge commented on Gruntfile.js:
```diff
@@ -3,194 +3,47 @@ module.exports = function (grunt) {
+        src: ['tests/dummy/app/templates/overview.hbs'],
+        overwrite: true,
+        replacements: [{
+          from: /The current version is .*\..*\..*\./,
```
Can we comment what this regex is matching?

Dylnuge commented on LICENSE.md:
```diff
@@ -0,0 +1,9 @@
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WI ...
```
Woah you changed the license. Can we even do this? Do Matt and Tom know about this? Please get approval from them.

azirbel replied:
WHOOPS didn't mean to. Will revert.

Dylnuge commented on README.md:
```diff
@@ -17,31 +17,65 @@ http://addepar.github.com/ember-table/
+
+Or, just include `dist/ember-table.js` and `dist/ember-table.css` in your app.
+
+Using Ember Table with bower is **deprecated** and will eventually be removed -
```
Full stop instead of a dash.

Dylnuge commented on README.md:
```diff
@@ -17,31 +17,65 @@ http://addepar.github.com/ember-table/
+
+Using Ember Table with bower is **deprecated** and will eventually be removed -
+we recommend that you migrate your apps to Ember CLI! Documentation has been
+updated to show Ember CLI usage; if you need documentation for globals-based
```
Full stop instead of a semicolon.

Dylnuge commented on README.md:
```diff
@@ -17,31 +17,65 @@ http://addepar.github.com/ember-table/
 http://emberjs.jsbin.com/xavipu/1/edit
 
-#### Installation
+#### Installation with Ember CLI
```
Maybe add a (Recommended) at the end of this heading to draw attention to it.

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+
+  // ---------------------------------------------------------------------------
+  // API - Inputs
+  // ---------------------------------------------------------------------------
```
Unrelated question: when does the attrs hash land in Ember? Thanks to clearly documented distinctions between our public and private APIs we should be in a pretty good place to move to that once it's ready.

azirbel replied:
Good question! I don't know and it's so hard to search for that kind of thing.

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    if (!Ember.$().mousewheel) {
+      throw 'Missing dependency: jquery-mousewheel';
+    }
+    if (!Ember.$().antiscroll) {
```
Are these the way to do dependency checking in ember-cli? I'm surprised by the use of the `Ember` global.

azirbel replied:
This shouldn't be needed in Ember CLI, since it's much harder to misconfigure when the install pulls in dependencies on its own. But it's still useful for the globals version, so I'll leave it for now.

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+  actions: {
+    addColumn: Ember.K,
+    sortByColumn: Ember.K
+  },
```
Conventionally even if it's empty I'd throw the actions hash at the bottom of the internals.

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    if (value) {
+      return value;
+    } else {
+      return Ember.A();
```
So getting it always returns a new empty array, and setting it returns the thing you set it to, but doesn't update anything? This code confuses me.

sreenath-addepar replied:
The return value is cached until it's invalidated (in this case never, since there are no dependent keys)

Dylnuge replied:
Ah yeah. Seems like a hacky way of relying on caching, but I don't think this is actually part of this PR, so my comment can be ignored for now.

cyril-sf replied:
`return Ember.A(value)` should be enough

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+  tableRowView: 'table-row',
+  tableRowViewClass: Ember.computed.alias('tableRowView'),
+
+  onColumnSort: function(column, newIndex) {
```
Looks like this is called by children (specifically headers). Should it be an action passed into those components instead of just a function?

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    }
+    var numFixedColumns = this.get('numFixedColumns') || 0;
+    return columns.slice(0, numFixedColumns) || [];
+  }).property('columns.@each', 'numFixedColumns'),
```
`columns.[]`

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    }
+    var numFixedColumns = this.get('numFixedColumns') || 0;
+    return columns.slice(numFixedColumns, columns.get('length')) || [];
+  }).property('columns.@each', 'numFixedColumns'),
```
`columns.[]`

Dylnuge commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+
+  onBodyContentLengthDidChange: Ember.observer(function() {
+    return Ember.run.next(this, function() {
+      return Ember.run.once(this, this.updateLayout);
```
A run once wrapped in a run next? What the what?

Dylnuge replied:
I feel like this should actually be a `scheduleOnce` on `afterRender`. I might be wrong though.

Dylnuge commented on addon/controllers/row-array.js:
```diff
@@ -0,0 +1,25 @@
+  content: null,
+
+  rowContent: Ember.computed(function() {
+    return [];
```
Elsewhere you use `Ember.A()`, why not here? (it makes a difference if prototype extensions are disabled)

Dylnuge commented on addon/views/column-sortable-indicator.js:
```diff
@@ -0,0 +1,12 @@
+export default Ember.View.extend(
+StyleBindingsMixin, RegisterTableComponentMixin, {
+  classNames: 'ember-table-column-sortable-indicator',
+  classNameBindings: 'tableComponent._isShowingSortableIndicator:active',
```
Should this reference a private (leading underscore) variable?

Dylnuge commented on addon/views/header-cell.js:
```diff
@@ -0,0 +1,148 @@
+  // jQuery UI resizable option
+  resizableOption: Ember.computed(function() {
+    return {
+      handles: 'e',
```
e?

azirbel replied:
means the "east" / "right" handle. Added comment

Dylnuge commented on addon/views/multi-item-collection.js:
```diff
@@ -0,0 +1,23 @@
+    var itemViewClassField = this.get('itemViewClassField');
+    var itemViewClass = attrs.content.get(itemViewClassField);
+    if (typeof itemViewClass === 'string') {
+      if (/[A-Z]+/.exec(itemViewClass)) {
```
This regex confuses me. Why the `+` if you're only checking if the first letter is capital?

Dylnuge commented on blueprints/ember-table/index.js:
```diff
@@ -0,0 +1,23 @@
+    // We assume that handlebars, ember, and jquery already exist
+    return this.addBowerPackagesToProject([
+      {
+        // Antiscroll seems to be abandoned by its original authors. We need
```
Oh no! Should we switch to something else at some point?

azirbel replied:
We may ditch antiscroll eventually...

Dylnuge commented on tests/dummy/app/templates/overview.hbs:
```diff
@@ -113,10 +115,9 @@
-$ grunt
-$ node examples.js</code></pre>
-      <p>Go to your browser and navigate to <a target="_BLANK" href="http://localhost:8000/gh_pages">localhost:8000/gh_pages</a></p>
+$ npm install && bower install
```
I prefer not putting chained instructions (using `&&` or `;`) in our docs. It's more straightforward to put each instruction on it's own line. Why is this `npm install && bower install` instead of (for instance) `npm install -g grunt-cli && npm install && bower install && ember serve`?

azirbel replied:
Fixed here and in the README.

Dylnuge commented:
I tried to limit my comments to actual changes you made (e.g. javascript conversion and ember-cli-ification) vs stuff that was preexisting, but I definitely slipped up in some places and you can ignore those comments.

Overall this LGTM (looks *great* to me). I'd really appreciate it if someone else glanced this over too.

Awesome work! Thanks @azirbel

azirbel commented on .editorconfig:
```diff
@@ -0,0 +1,33 @@
+indent_size = 2
+
+[*.{diff,md}]
+trim_trailing_whitespace = false
```
dunno! grabbed from @igillis - got opinions on these two comments, Ian?

igillis replied:
agree -- kill the repeated rules and trim whitespace in markdown

bryanaka replied:
you do not want to trim whitespace in markdown, since two trailing spaces '  ' means a true line break. But yes, the rest do not need to be repeated

cyril-sf commented on README.md:
````diff
@@ -17,31 +17,65 @@ http://addepar.github.com/ember-table/
+
+```bash
+$ npm install -g grunt-cli
+$ npm install && bower install
````
we should just make `bower install` a part of the `postinstall` steps of `npm`.

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+          this.get('persistedSelection').addObject(val);
+          break;
+        case 'multiple':
+          this.get('persistedSelection').addObjects(val);
```
probably worth adding `break` even if there are no other cases

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+  },
+
+  // An array of Ember.Table.Row computed based on `content`
+  bodyContent: Ember.computed(function() {
```
This is probably out of scope, but we should not need to recreate the controller each time the content changes

```js
bodyContent: Ember.computed(function() {
    return RowArrayController.create({
      target: this,
      parentController: this,
      container: this.get('container'),
      itemController: Row,
      content: Ember.computed.alias 'parentController.content'
    });
  }).property()
```

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+  }).property('columns.@each', 'numFixedColumns'),
+
+  prepareTableColumns: function() {
+    var _this = this;
```
`var emberTable = this`?

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+
+  getNextResizableColumn: function(columns, index) {
+    var column;
+    while (index < columns.length) {
```
Can we rely on `Ember.Array.findBy`?

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    // window. Otherwise, we get this warning when used in tests. You have
+    // turned on testing mode, which disabled the run-loop's autorun. You
+    // will need to wrap any code with asynchronous side-effects in an
+    // Ember.run
```
We probably need to follow that wise comment at some point :stuck_out_tongue:

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    if ((this.get('_state') || this.get('state')) !== 'inDOM') {
+      return;
+    }
+    this.set('_width', this.$().parent().width());
```
I would store `this.$().parent()` in a variable

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+  },
+
+  elementSizeDidChange: function() {
+    if ((this.get('_state') || this.get('state')) !== 'inDOM') {
```
We probably should have a helper for that

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    if (this.tableWidthNowTooSmall()) {
+      this.set('columnsFillTable', true);
+    }
+    return Ember.run(this, this.elementSizeDidChange);
```
`return` seems to be coming from the CoffeeScript to JavaScript conversion

azirbel replied:
thanks for catching these!

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+
+  elementSizeDidChange: function() {
+    if ((this.get('_state') || this.get('state')) !== 'inDOM') {
+      return;
```
`return` seems to be coming from the CoffeeScript to JavaScript conversion

Dylnuge replied:
Seems valid here, since it short circuits the rest of the logic. Might be better to invert the logic and wrap the rest of the function it though.

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    this.set('_height', this.$().parent().height());
+    // we need to wait for the table to be fully rendered before antiscroll can
+    // be used
+    return Ember.run.next(this, this.updateLayout);
```
`return` seems to be coming from the CoffeeScript to JavaScript conversion

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    // updating antiscroll
+    this.$('.antiscroll-wrap').antiscroll().data('antiscroll').rebuild();
+    if (this.get('columnsFillTable')) {
+      return this.doForceFillColumns();
```
`return` seems to be coming from the CoffeeScript to JavaScript conversion

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+      bodyHeight -= this.get('footerHeight');
+    }
+    return bodyHeight;
+  }).property('_tablesContainerHeight', '_hasHorizontalScrollbar',
```
`_hasHorizontalScrollbar` is not used in the function, not sure if it's needed here

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    return this.get('rowHeight') * this.get('bodyContent.length');
+  }).property('rowHeight', 'bodyContent.length'),
+
+  _tableContainerWidth: Ember.computed(function() {
```
`Ember.computed.oneWay`

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+    if (columnWidthPath == null) {
+      columnWidthPath = 'width';
+    }
+    if (!columns) {
```
small optimization, but that should be at the top of the function

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+
+  lastSelected: null,
+
+  isSelected: function(row) {
```
Shouldn't that be a CP?

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+            lastIndex = 0;
+          }
+
+          var curIndex = this.rowIndex(this.getRowForEvent(event));
```
`this.getRowForEvent(event)` has already been store at the beginning in `row`

cyril-sf commented on addon/components/ember-table.js:
```diff
@@ -0,0 +1,564 @@
+
+          this.get('rangeSelection').addObjects(
+            this.get('bodyContent').slice(minIndex, maxIndex + 1)
+            .mapBy('content'));
```
I would close the parenthesis on the next line.

cyril-sf commented on addon/mixins/style-bindings.js:
```diff
@@ -1,6 +1,7 @@
-Ember.AddeparMixins = Ember.AddeparMixins || Ember.Namespace.create();
+// TODO(azirbel): This needs to be an external dependency.
```
Is this TODO relevant?

azirbel replied:
yes, it's referring to this file (style bindings) and not to Ember. I'll add a line break. We want to share `style-bindings.js` between our open source repositories

cyril-sf commented on addon/mixins/style-bindings.js:
```diff
@@ -15,11 +16,10 @@ Ember.AddeparMixins.StyleBindingsMixin = Ember.Mixin.create({
   },
-  applyStyleBindings: Ember.on('init',
-    Ember.observer('styleBindings', function() {
+  applyStyleBindings: function() {
```
Why not keeping `Ember.on('init', `?

cyril-sf commented on addon/mixins/style-bindings.js:
```diff
@@ -15,11 +16,10 @@ Ember.AddeparMixins.StyleBindingsMixin = Ember.Mixin.create({
     var lookup, properties, styleBindings, styleComputed, styles,
       _this = this;
-    styleBindings = this.get('styleBindings');
+    styleBindings = this.styleBindings;
```
why not keeping `this.get`?

azirbel replied:
not sure what happened here. Changed back.

cyril-sf commented on addon/views/header-cell.js:
```diff
@@ -0,0 +1,148 @@
+    var nextColumnMaxDiff = this.get('nextResizableColumn.maxWidth') -
+        this.get('nextResizableColumn.width');
+    if (this.get('minWidth') && nextColumnMaxDiff) {
+      return Math.min(this.get('minWidth'), this.get('width') -
```
I usually prefer to store the return value in one variable and have a single `return` statement when possible. It's easier to read IMO.

cyril-sf commented on addon/views/header-cell.js:
```diff
@@ -0,0 +1,148 @@
+  }).property('effectiveMinWidth', 'effectiveMaxWidth'),
+
+  didInsertElement: function() {
+    this.elementSizeDidChange();
```
We probably want to call `_super` at some point

cyril-sf commented on addon/views/header-cell.js:
```diff
@@ -0,0 +1,148 @@
+  },
+
+  cellWidthDidChange: Ember.observer(function() {
+    return Ember.run.schedule('afterRender', this, this.elementSizeDidChange);
```
`return` is not needed

cyril-sf commented on addon/views/header-cell.js:
```diff
@@ -0,0 +1,148 @@
+  }, 'width'),
+
+  resizableObserver: Ember.observer(function() {
+    return this.recomputeResizableHandle();
```
`return` is not needed

cyril-sf commented on addon/views/lazy-table-block.js:
```diff
@@ -0,0 +1,19 @@
+  scrollTop: null,
+
+  onScrollLeftDidChange: Ember.observer(function() {
+    return this.$().scrollLeft(this.get('scrollLeft'));
```
`return` is not needed

cyril-sf commented on addon/views/lazy-item.js:
```diff
@@ -0,0 +1,21 @@
+    return this.get('itemIndex') * this.get('rowHeight');
+  }).property('itemIndex', 'rowHeight'),
+
+  // TODO(azirbel): Add explicit else case
```
Based on this comment, did we return something in the `else` case in coffeescript?

azirbel replied:
no, this is the correct translated JS. Previous coffeescript was
```
display: Ember.computed ->
  'none' if not @get('content')
.property 'content'
```
I guess it implicitly returns `undefined`? Which seems pretty sloppy.

cyril-sf commented on addon/views/table-block.js:
```diff
@@ -0,0 +1,22 @@
+  scrollLeft: null,
+
+  onScrollLeftDidChange: Ember.observer(function() {
+    return this.$().scrollLeft(this.get('scrollLeft'));
```
`return` is not needed

cyril-sf commented on addon/views/table-cell.js:
```diff
@@ -0,0 +1,63 @@
+  init: function() {
+    this._super();
+    this.contentPathDidChange();
+    return this.contentDidChange();
```
`return` is not needed

cyril-sf commented on addon/views/table-cell.js:
```diff
@@ -0,0 +1,63 @@
+  width: Ember.computed.alias('column.width'),
+
+  contentDidChange: function() {
+    return this.notifyPropertyChange('cellContent');
```
`return` is not needed?

cyril-sf commented on addon/views/table-cell.js:
```diff
@@ -0,0 +1,63 @@
+  contentPathWillChange: Ember.beforeObserver(function() {
+    var contentPath = this.get('column.contentPath');
+    if (contentPath) {
+      return this.removeObserver("row." + contentPath, this,
```
`return` is not needed

cyril-sf commented on addon/views/table-cell.js:
```diff
@@ -0,0 +1,63 @@
+  contentPathDidChange: Ember.beforeObserver(function() {
+    var contentPath = this.get('column.contentPath');
+    if (contentPath) {
+      return this.addObserver("row." + contentPath, this,
```
`return` is not needed

cyril-sf commented on package.json:
```diff
@@ -1,34 +1,70 @@
+    "node": ">= 0.10.0"
+  },
+  "author": "",
+  "license": "BSD",
```
`MIT`?

cyril-sf commented on CONTRIBUTING.md:
```diff
@@ -139,7 +137,7 @@ included in the project:
 
 **IMPORTANT**: By submitting a patch, you agree to allow the project owners to
-license your work under the terms of the [BSD License](LICENSE).
+license your work under the terms of the [BSD License](LICENSE.md).
```
`MIT`?

azirbel replied:
It's actually BSD, I changed the LICENSE file below by mistake but I'm changing it back

cyril-sf commented on CONTRIBUTING.md:
```diff
@@ -158,7 +156,8 @@ Coming soon!
 
 By contributing your code, you agree to license your contribution under the
-[BSD license](LICENSE).
+[BSD license](LICENSE.md).
```
`MIT`?

cyril-sf commented on package.json:
```diff
@@ -1,34 +1,70 @@
+    "broccoli-es6modules": "~0.5.1"
+  },
+  "keywords": [
+    "ember-addon"
```
maybe we want to add `ember-table` or something similar

cyril-sf commented on tests/dummy/app/components/configurable-table.js:
```diff
@@ -0,0 +1,13 @@
+// when its width has changed.
+export default TableComponent.extend({
+  layoutName: 'components/ember-table',
+  parentWidthObserver: Ember.computed(function() {
```
Weird name as it's not an observer

azirbel replied:
uhm humm. It is supposed to be an observer! Weird... https://github.com/Addepar/ember-table/blob/724cb261c45b87344d20c7d814a6622f3621ab44/app/controllers/configurable-columns.coffee#L131

cyril-sf commented on tests/dummy/app/components/financial-table.js:
```diff
@@ -0,0 +1,191 @@
+        return _this.createTree(row, child);
+      };
+    })(this));
+    // TODO(Peter): Hack... only collapse table if it should collapseByDefault
```
This is sad that we have comments like that in our demo app...

cyril-sf commented on tests/dummy/app/components/financial-table.js:
```diff
@@ -0,0 +1,191 @@
+
+  computeStyles: function(parent, node) {
+    node.computeStyles(parent);
+    return node.get('children').forEach((function(_this) {
```
`return` statements are not needed.

cyril-sf commented on tests/dummy/app/components/financial-table.js:
```diff
@@ -0,0 +1,191 @@
+        return;
+      }
+      children.forEach(function(child) {
+        return child.recursiveCollapse(isCollapsed);
```
`return` is not needed

cyril-sf commented on tests/dummy/app/components/financial-table.js:
```diff
@@ -0,0 +1,191 @@
+      children.forEach(function(child) {
+        return child.recursiveCollapse(isCollapsed);
+      });
+      return this.notifyPropertyChange('rows');
```
`return` is not needed

cyril-sf commented on tests/dummy/app/components/financial-table.js:
```diff
@@ -0,0 +1,191 @@
+    toggleCollapse: function(row) {
+      row.toggleProperty('isCollapsed');
+      return Ember.run.next(this, function() {
+        return this.notifyPropertyChange('rows');
```
`return` is not needed

cyril-sf commented on tests/dummy/app/components/financial-table.js:
```diff
@@ -0,0 +1,191 @@
+
+    toggleCollapse: function(row) {
+      row.toggleProperty('isCollapsed');
+      return Ember.run.next(this, function() {
```
`return` is not needed

cyril-sf commented on tests/dummy/app/controllers/ajax.js:
```diff
@@ -0,0 +1,31 @@
+
+  tableContent: Ember.computed(function() {
+    return AjaxTableLazyDataSource.create({
+      content: new Array(100)
```
out of scope, but this should be used through a variable

cyril-sf commented on tests/dummy/app/controllers/configurable-columns.js:
```diff
@@ -0,0 +1,89 @@
+  }.property('isFluid'),
+
+  updateDemoTableWidth: function(newWidth) {
+    return this.set('demoTableWidth', newWidth);
```
`return` is not needed

cyril-sf commented on tests/dummy/app/controllers/dynamic-bars.js:
```diff
@@ -0,0 +1,57 @@
+
+export default Ember.Controller.extend({
+  // TODO(azirbel): Don't use setInterval in an Ember application
+  init: function() {
```
probably should call `_super`

cyril-sf commented on tests/dummy/app/controllers/dynamic-bars.js:
```diff
@@ -0,0 +1,57 @@
+  // TODO(azirbel): Don't use setInterval in an Ember application
+  init: function() {
+    var _this = this;
+    setInterval(function() {
```
out of scope, but this should be canceled when the controller is torn down

cyril-sf commented on tests/dummy/app/controllers/dynamic-bars.js:
```diff
@@ -0,0 +1,57 @@
+  init: function() {
+    var _this = this;
+    setInterval(function() {
+      return _this.get('tableContent').forEach(function(item) {
```
`return` is not needed

cyril-sf commented on tests/dummy/app/controllers/dynamic-bars.js:
```diff
@@ -0,0 +1,57 @@
+    return columns;
+  }),
+
+  tableContent: Ember.computed(function() {
```
code duplication with `BarsController`

cyril-sf commented on tests/dummy/app/controllers/sparkline.js:
```diff
@@ -0,0 +1,71 @@
+
+export default Ember.Controller.extend({
+  tableColumns: Ember.computed(function() {
+    var name = ColumnDefinition.create({
```
out of scope, but I would create an array of column definitions, and only after use `ColumnDefintion.create`

```js
var columnDefintions = [{
    ...
}, {
    ...
}]

return columnDefintions.map( function( columnDefintion ) {
   return ColumnDefinition.create( columnDefintion );
});
```

cyril-sf commented on tests/dummy/app/routes/overview.js:
```diff
@@ -0,0 +1,14 @@
+export default Ember.Route.extend({
+  activate: function() {
+    var controller = this.controllerFor('application');
+    return controller.set('showLargeHero', true);
```
`return` not needed

cyril-sf commented on tests/dummy/app/routes/overview.js:
```diff
@@ -0,0 +1,14 @@
+
+  deactivate: function() {
+    var controller = this.controllerFor('application');
+    return controller.set('showLargeHero', false);
```
`return` not needed

cyril-sf commented on tests/dummy/app/templates/bars.hbs:
```diff
@@ -0,0 +1,41 @@
+    <div class="col-md-12 bumper-30">
+      <h3>Controller</h3>
+      <div class="highlight">
+        <pre class="prettyprint lang-coffee">@@{controllers/bars.js}</pre>
```
`lang-js` ( there are other occurences )

cyril-sf commented on tests/dummy/app/templates/configurable-columns.hbs:
```diff
@@ -70,9 +69,9 @@
+      <h3>Template</h3>
       <div class="highlight">
-<pre class="prettyprint lang-html">&#123;&#123;table-component
+<pre class="lang-html">&#123;&#123;table-component
```
do we really want to remove `prettyprint`?

azirbel replied:
restored and rewrote this logic

cyril-sf commented on tests/dummy/app/templates/configurable-columns.hbs:
```diff
@@ -82,14 +81,22 @@
+      <h3>Controller</h3>
       <div class="highlight">
-<pre>App.ApplicationController = Ember.Controller.extend
+<pre class="lang-coffee">App.ApplicationController = Ember.Controller.extend
```
We probably want to update this documentation to use the ES6 module syntax. This is not relevant for an Ember-CLI application.

azirbel replied:
great catch, thanks! Updated the docs.

cyril-sf commented on tests/dummy/app/templates/migration-guides.hbs:
```diff
@@ -67,7 +67,7 @@
         Version 0.3.0 includes major column resizing changes, but with
         <code>canAutoResize</code> defaulting to <code>true</code>.  We
-        recommend skipping this version and upgrading directly to 0.4.0; see
+        recommend skipping this version and upgrading direcly to 0.4.0; see
```
added typo ^_^

azirbel replied:
I ... don't know how I managed this

Dylnuge replied:
You use vim right? You probably just hit x by mistake.
On May 20, 2015 11:45 PM, "Alex Zirbel" <notifications@github.com> wrote:

> In tests/dummy/app/templates/migration-guides.hbs
> <https://github.com/Addepar/ember-table/pull/309#discussion_r30776469>:
>
> > @@ -67,7 +67,7 @@
> >        <p>
> >          Version 0.3.0 includes major column resizing changes, but with
> >          <code>canAutoResize</code> defaulting to <code>true</code>.  We
> > -        recommend skipping this version and upgrading directly to 0.4.0; see
> > +        recommend skipping this version and upgrading direcly to 0.4.0; see
>
> I ... don't know how I managed this
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Addepar/ember-table/pull/309/files#r30776469>.
>

azirbel replied:
No, no, vim couldn't have been responsible for this mistake. Maybe I was trying out emacs at the time.

cyril-sf commented on tests/dummy/app/templates/overview.hbs:
```diff
@@ -35,7 +37,7 @@
         </li>
         <li><a target="_BLANK" href="https://github.com/brandonaaron/jquery-mousewheel">jquery.mousewheel</a></li>
-        <li><a target="_BLANK" href="https://github.com/LearnBoost/antiscroll">antiscroll</a></li>
+        <li><a target="_BLANK" href="https://github.com/azirbel/antiscroll">antiscroll</a></li>
```
Addepar should host that repo.

cyril-sf replied:
Also it seems like the ownership of the project has been transferred https://github.com/Automattic/antiscroll

Unless we have custom modifications that are not accepted through a PR, we probably want to rely on the "official" one.

azirbel replied:
We do have custom modifications, and even though the project is transferred, PRs are being ignored. We can move this to an Addepar repository, though. I'll add that to my list and do it as a separate project.

cyril-sf commented on tests/dummy/app/templates/simple.hbs:
```diff
@@ -0,0 +1,28 @@
+      </div>
+    </div>
+
+    {{!-- TODO(azirbel): Re-enable prettyprint or similar formatter --}}
```
Is this comment still valid?

azirbel replied:
yep, prettyprint doesn't seem happy but it's pretty minor

cyril-sf commented on tests/dummy/app/utils/number-format.js:
```diff
@@ -0,0 +1,20 @@
+// HACK: Used to help format table cells, should be refactored or use a library
+// TODO(azirbel): Should be a handlebars helper
```
+1

cyril-sf commented on tests/dummy/app/views/ajax-table-lazy-data-source.js:
```diff
@@ -0,0 +1,43 @@
+        return _this.createGithubEvent(row, event);
+      });
+    });
+    return _.range(start, end).forEach(function(index) {
```
`return` statements not needed

cyril-sf commented on tests/dummy/app/views/configurable-column-definition.js:
```diff
@@ -0,0 +1,57 @@
+  }).property('maxWidth'),
+
+  headerCellNameLowerCase: Ember.computed(function() {
+    return this.get('headerCellName').toLowerCase();
```
that also look like a handlebars helper

cyril-sf commented on tests/dummy/app/views/editable-table-cell.js:
```diff
@@ -0,0 +1,29 @@
+    typeBinding: 'parentView.type',
+    valueBinding: 'parentView.cellContent',
+    didInsertElement: function() {
+      return this.$().focus();
```
`return` not needed

We also need to call `_super`

cyril-sf commented on tests/dummy/app/views/editable-table-cell.js:
```diff
@@ -0,0 +1,29 @@
+      return this.$().focus();
+    },
+    focusOut: function() {
+      return this.set('parentView.isEditing', false);
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/editable-table-cell.js:
```diff
@@ -0,0 +1,29 @@
+  }),
+
+  onRowContentDidChange: Ember.observer(function() {
+    return this.set('isEditing', false);
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/editable-table-cell.js:
```diff
@@ -0,0 +1,29 @@
+
+  onRowContentDidChange: Ember.observer(function() {
+    return this.set('isEditing', false);
+  }).observes('row.content'),
```
This is not the correct syntax for an observer, it mixes the 2 syntaxes ( with and without prototype extension )

cyril-sf commented on tests/dummy/app/views/editable-table-cell.js:
```diff
@@ -0,0 +1,29 @@
+
+  click: function(event) {
+    this.set('isEditing', true);
+    return event.stopPropagation();
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/financial-table-tree-row.js:
```diff
@@ -0,0 +1,88 @@
+    }
+    this.set('groupingLevel', groupingLevel);
+    this.set('indentation', indentation);
+    return this.set('isShowing', isShowing);
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/financial-table-tree-row.js:
```diff
@@ -0,0 +1,88 @@
+  computeRowStyle: function(maxLevels) {
+    var level;
+    level = this.getFormattingLevel(this.get('groupingLevel'), maxLevels);
+    return this.set('rowStyle', 'ember-table-row-style-' + level);
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/financial-table-tree-row.js:
```diff
@@ -0,0 +1,88 @@
+
+  recursiveCollapse: function(isCollapsed) {
+    this.set('isCollapsed', isCollapsed);
+    return this.get('children').forEach(function(child) {
```
`return` statements not needed

cyril-sf commented on tests/dummy/app/views/horizon-table-cell.js:
```diff
@@ -0,0 +1,35 @@
+
+  onWidthDidChange: Ember.observer(function() {
+    this.$('svg').remove();
+    return this.renderD3View();
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/horizon-table-cell.js:
```diff
@@ -0,0 +1,35 @@
+  onWidthDidChange: Ember.observer(function() {
+    this.$('svg').remove();
+    return this.renderD3View();
+  }).observes('width'),
```
wrong syntax for an observer

cyril-sf commented on tests/dummy/app/views/horizon-table-cell.js:
```diff
@@ -0,0 +1,35 @@
+  }).observes('width'),
+
+  didInsertElement: function() {
+    return this.onWidthDidChange();
```
`return` not needed

missing `_super`

cyril-sf commented on tests/dummy/app/views/horizon-table-cell.js:
```diff
@@ -0,0 +1,35 @@
+    data = this.get('horizonContent');
+    chart = d3HorizonUtils.d3Horizon().width(width).height(height).bands(2).mode('mirror').interpolate('basis');
+    svg = d3.select('#' + this.get('elementId')).append('svg').attr('width', width).attr('height', height);
+    return svg.data([data]).call(chart);
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/rating-table-cell.js:
```diff
@@ -0,0 +1,31 @@
+  templateName: 'editable-table/rating-table-cell',
+
+  onRowContentDidChange: Ember.observer(function() {
+    return this.applyRating(this.get('cellContent'));
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/rating-table-cell.js:
```diff
@@ -0,0 +1,31 @@
+
+  onRowContentDidChange: Ember.observer(function() {
+    return this.applyRating(this.get('cellContent'));
+  }).observes('cellContent'),
```
wrong syntax for an observer

cyril-sf commented on tests/dummy/app/views/rating-table-cell.js:
```diff
@@ -0,0 +1,31 @@
+
+  didInsertElement: function() {
+    this._super();
+    return this.onRowContentDidChange();
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/rating-table-cell.js:
```diff
@@ -0,0 +1,31 @@
+  applyRating: function(rating) {
+    this.$('.rating span').removeClass('active');
+    var span = this.$('.rating span').get(rating);
+    return Ember.$(span).addClass('active');
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/rating-table-cell.js:
```diff
@@ -0,0 +1,31 @@
+      return;
+    }
+    this.get('column').setCellContent(this.get('row'), rating);
+    return this.applyRating(rating);
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/sparkline-table-cell.js:
```diff
@@ -0,0 +1,42 @@
+
+  onContentOrSizeDidChange: Ember.observer(function() {
+    this.$('svg').remove();
+    return this.renderD3View();
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/sparkline-table-cell.js:
```diff
@@ -0,0 +1,42 @@
+  onContentOrSizeDidChange: Ember.observer(function() {
+    this.$('svg').remove();
+    return this.renderD3View();
+  }).observes('row', 'width'),
```
wrong syntax for an observer ( I'll stop commenting on this one as you can search the occurrences in the repo )

cyril-sf commented on tests/dummy/app/views/sparkline-table-cell.js:
```diff
@@ -0,0 +1,42 @@
+  }).observes('row', 'width'),
+
+  didInsertElement: function() {
+    return this.renderD3View();
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/sparkline-table-cell.js:
```diff
@@ -0,0 +1,42 @@
+    });
+    var svg = d3.select('#' + (this.get('elementId'))).append('svg:svg').attr('height', h).attr('width', w);
+    var g = svg.append('svg:g');
+    return g.append('svg:path').attr('d', line(data)).attr('stroke', function() {
```
`return` not needed

cyril-sf commented on tests/dummy/app/views/sparkline-table-cell.js:
```diff
@@ -0,0 +1,42 @@
+    var svg = d3.select('#' + (this.get('elementId'))).append('svg:svg').attr('height', h).attr('width', w);
+    var g = svg.append('svg:g');
+    return g.append('svg:path').attr('d', line(data)).attr('stroke', function() {
+      return fill(Math.round(Math.random()) * 10);
```
`return` not needed?

cyril-sf commented on tests/dummy/config/environment.js:
```diff
@@ -0,0 +1,67 @@
+    // you try to load a route directly, you'll get a 404.)
+    locationType: 'hash',
+    EmberENV: {
+      // TODO(azirbel): Disable prototype extensions
```
We observes arrays, not sure if we can do that afterall

cyril-sf commented:
That's a tremendous effort, I'm super excited by this PR!

Some of my comments are probably out of scope, let me know when you think it's the case.

azirbel commented:
Self-reminder to change the required version of ember to `>=1.4 <1.9`

azirbel commented:
Thanks for the review @Dylnuge and @cyril-sf! I've addressed all your comments (ignoring all the ones for code I just moved). It was especially good that you guys cought a lot of `return` statements, bad `observes` syntax, and the ninja accidental license change `>.>`

I think this is good to merge, and Dylan has LGTM'd it. Let me know if you have objections.

azirbel commented:
:tada:

ldevalliere commented:
:heart:

SirZach commented:
congrats!

cyril-sf commented:
:fireworks:

artych commented:
great!! 
waiting for ember-widget-addon
